File: server/spec/factories/git/repositories.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :git_repository, class: 'Devops::GitRepository' do
    association :credential, factory: :git_provider_credential
    association :account

    sequence(:external_id) { |n| "repo#{n}" }
    sequence(:name) { |n| "test-repo-#{n}" }
    sequence(:full_name) { |n| "testuser/test-repo-#{n}" }
    owner { 'testuser' }
    description { 'A test repository' }
    default_branch { 'main' }
    clone_url { "https://github.com/#{owner}/#{name}.git" }
    ssh_url { "git@github.com:#{owner}/#{name}.git" }
    web_url { "https://github.com/#{owner}/#{name}" }
    is_private { false }
    is_fork { false }
    is_archived { false }
    webhook_configured { false }
    webhook_id { nil }
    webhook_secret { nil }
    languages { { 'Ruby' => 70, 'JavaScript' => 30 } }
    topics { %w[rails api] }
    stars_count { 10 }
    forks_count { 2 }
    open_issues_count { 5 }
    open_prs_count { 1 }
    last_synced_at { nil }
    last_commit_at { nil }

    trait :private do
      is_private { true }
    end

    trait :fork do
      is_fork { true }
    end

    trait :archived do
      is_archived { true }
    end

    trait :with_webhook do
      webhook_configured { true }
      webhook_id { 'webhook_123' }
      webhook_secret { SecureRandom.hex(20) }
    end

    trait :synced do
      last_synced_at { 1.hour.ago }
      last_commit_at { 2.hours.ago }
    end

    trait :popular do
      stars_count { 1000 }
      forks_count { 100 }
    end
  end
end

File: frontend/src/features/devops/git/types/repositories.ts
// Git Repository Types

import type { PipelineStats } from './pipelines';

export interface PaginationInfo {
  current_page: number;
  per_page: number;
  total_pages: number;
  total_count: number;
}

export type BranchFilterType = 'none' | 'exact' | 'wildcard' | 'regex';

export interface GitRepository {
  id: string;
  name: string;
  full_name: string;
  owner: string;
  description?: string;
  default_branch: string;
  web_url?: string;
  is_private: boolean;
  is_fork: boolean;
  is_archived: boolean;
  webhook_configured: boolean;
  branch_filter?: string;
  branch_filter_type?: BranchFilterType;
  branch_filter_enabled?: boolean;
  stars_count: number;
  forks_count: number;
  open_issues_count: number;
  open_prs_count: number;
  primary_language?: string;
  topics: string[];
  last_synced_at?: string;
  last_commit_at?: string;
  created_at: string;
  provider_type: string;
  credential_id: string;
}

export interface GitRepositoryDetail extends GitRepository {
  clone_url?: string;
  ssh_url?: string;
  languages?: Record<string, number>;
  sync_settings?: Record<string, unknown>;
  webhook_id?: string;
  provider_created_at?: string;
  provider_updated_at?: string;
  pipeline_stats?: PipelineStats;
  credential: {
    id: string;
    name: string;
    provider_name: string;
  };
}

// Available repository from provider (not yet imported)
export interface AvailableRepository {
  external_id: string;
  name: string;
  full_name: string;
  owner: string;
  description?: string;
  default_branch: string;
  web_url?: string;
  is_private: boolean;
  is_fork: boolean;
  is_archived: boolean;
  stars_count: number;
  primary_language?: string;
  already_imported: boolean;
}

export interface RepositoryUsage {
  current: number;
  limit: number;
  available: number;
}

export interface AvailableRepositoriesResponse {
  repositories: AvailableRepository[];
  pagination: {
    current_page: number;
    per_page: number;
    total_count: number;
  };
  usage: RepositoryUsage;
}

export interface ImportRepositoriesResult {
  imported_count: number;
  error_count: number;
  repositories: GitRepository[];
  errors: Array<{
    external_id: string;
    error: string;
  }>;
  usage: {
    current: number;
    limit: number;
  };
  message: string;
}

export interface GitRepositoriesResponse {
  repositories: GitRepository[];
  pagination: PaginationInfo;
}

export interface WebhookConfigFormData {
  branch_filter?: string;
  branch_filter_type?: BranchFilterType;
}
